src/parquet/bloom_filter_algorithm.h

```diff
@@ -0,0 +1,87 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_BLOOM_FILTER_ALGORITHM_H
+#define PARQUET_BLOOM_FILTER_ALGORITHM_H
+
+#include <cstdint>
+#include "parquet/types.h"
+
+namespace parquet {
+
+// Abstract class for Bloom filter algorithm
+class BloomFilterAlgorithm {
```

Comment:
Sorry, I don't quite understand this.  Why do we need to implement data structure polymorphism other than simple algorithm polymorphsim?  From Bloom filters perspective, what it needs is testBits and setBits, abstract these two interfaces should open maximum capability to the implementer. We may not know how the data structure be designed for other algorithms.

Reply:
Abstracting at the data structure level provides much more flexibility.

Consider how you do things now: InitBitset is defined in the BloomFilter class, but that might be different for different algorithms. The way it is written now actually constrains the data structure more tightly.

Reply:
OK. Do you mean we should move bitset and its intializer to BloomFilterAlgorithm class and then put specific initializer to implementer such as BlockBloomFilterAlgorithm?

Reply:
I moved bitset and related operations to algorithm class. Since the bitset we get from inputstream maybe not 64-byte aligned and no way to cast it to uint32_t pointer, so I have to use memcpy here.